### PR TITLE
httpload: write an HTTP load testing program

### PR DESCRIPTION
**Pull request by rgianassi**

### CODE REVIEW

File: url_shortener/cmd/httpload/main.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+func computeTimings() (timings []float64, err error) {
+	timings = make([]float64, 0, 512)
+
+	timings = append(timings, 0.0003)
+	timings = append(timings, 0.0054)
+	timings = append(timings, 0.0024)
+	timings = append(timings, 0.0295)
+
+	return timings, err
```
Comment (guerinoni) [inline url_shortener/cmd/httpload/main.go:53]: `err` is not used!

File: url_shortener/cmd/httpload/main.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+	exitCodeError = 1
+)
+
+func computeStatuses() (statuses map[int]int, err error) {
+	statuses = make(map[int]int)
+
+	statuses[200] = 198
+	statuses[404] = 2
+
+	return statuses, err
```
Comment (guerinoni) [inline url_shortener/cmd/httpload/main.go:22]: `err` is never used!

**Comment by rgianassi** [comment]
Thank you @guerinoni for spending time reviewing my bad code! :smile: 
I added some comments.

File: url_shortener/cmd/httpload/main.go (hunk 2 above)
Comment (rgianassi) [reply url_shortener/cmd/httpload/main.go:22]: Nitpicking: `err` is used in `main` function.

At the moment, I only developed the print statistics part and I am still unsure that `computeStatuses` will be part of the final API.
However, if `computeStatuses` is going to be the final API it has to return errors in case of problems, so I defined the `err` variable. This has the nice side effect that the `main` function is already "ready for production".
Currently, `computeStatuses` has a fake implementation, so it cannot make "mistakes": this is the reason is always returning a nil `err` value. For the moment. :smile:

File: url_shortener/cmd/httpload/main.go (hunk 1 above)
Comment (rgianassi) [reply url_shortener/cmd/httpload/main.go:53]: Same comment as of `computeStatuses` applies.

File: url_shortener/cmd/httpload/main.go [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+	c := make(chan error, 1)
+	req = req.WithContext(ctx)
+	go func() { c <- f(http.DefaultClient.Do(req)) }()
+	select {
+	case <-ctx.Done():
+		<-c // Wait for f to return.
+		return ctx.Err()
+	case err := <-c:
+		return err
+	}
```
Comment (arl) [inline url_shortener/cmd/httpload/main.go:135]: If the context you pass to the request (as you do in `WithContext`) gets cancelled before the end of the request, the request will fail (`Do() returns an error`).
Now, to know if the request failed because of the context you have to analyze the error returned by Do.
See an example here: https://play.golang.org/p/RI9TTyvz0xO

Calling `ctx.Done()` yourself is fragile for that purpose, since the context might expire between the call to `DefaultClient.Do` and the `<-ctx.Done`

**Review by arl — COMMENTED** [review]
It would be great if you could structure the program as follows:
 - first a `Config struct` which holds all the configuration parameters at once. You'd have a kind of constructor function such as `newConfigFromFlags(flags *flag.FlagSet)(Config, error)` that returns does the parsing, then fills and returns a Config instance. In this way you could easily write unit-tests of the parsing and verifies constraints such as _parameter a and b are exclusive_ and returns an error.
 - second, have another struct called `loadTester` or something which you create by passing a Config instance.

This is not very different to what you did, it's just that it's more idiomatic, so you're going to see this many times in Go. Generally, constructor-like functions do not take more than 2 or 3 parameters, if they're more then using a Config structure might be a good idea. Also it's more easier to test since you can create `flag.FlagSet` in unit-tests.


Another idea, since the flag parsing si relatively trivial, you could just parse as you did but fill a Config structure directly with the flag values, then have a `Config.Check() error` method (very similar to your `checkFlags` function) which verifies the cli parameter match the expected rules.

File: url_shortener/cmd/httpload/main.go [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+
+	totalTiming := float64(0)
+	slowestTiming := float64(0)
+	fastestTiming := float64(0)
+	averageTiming := float64(0)
+	requestsPerSecond := float64(0)
+
+	n := len(results)
+
+	if n > 0 {
```
[the remark this answers is not in the document]
Comment (rgianassi) [reply url_shortener/cmd/httpload/main.go:78]: In that case I have to double the `Fprintf` statements. Or do you have an idea to print zero or one value?

File: url_shortener/cmd/httpload/main.go [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+		for i := 1; i < n; i++ {
+			timing := float64(results[i].timing) * nanoSecondRatio
+
+			totalTiming += timing
+
+			// the following conditions are counter-intuitive,
+			// but a bigger timing means a slower execution and
+			// a tinier timing means a faster execution
+			// this is the reason I prefer to measure speed
+			// instead of time :)
```
[the remark this answers is not in the document]
Comment (rgianassi) [reply url_shortener/cmd/httpload/main.go:93]: Timing <=> Duration, but I can change names if you prefer.

File: url_shortener/cmd/httpload/main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (arl) [reply url_shortener/cmd/httpload/main.go:78]: I'd say that in the case there are no results, it'd be preferable for `dumpTimings` to either do nothing or return an error.
Also, to keep the scope of the function as general as possible, it'd nice that dumpTimings takes an `io.Writer` in which you write the timings, rather than returning a string.

By doing so, you let to the caller the choice of the destination, it could be `os.Stdout`, `os.Stderr` or even a http.Response or whatever. 
I know the caller can achieve the same thing with a `string`, but trying to think in terms of streaming interfaces (`io.Writer`, `io.Reader`) is good as you're going to see a lot of that in Go

File: url_shortener/cmd/httpload/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (arl) [reply url_shortener/cmd/httpload/main.go:93]: no no timings is ok. The sentence about _speed_ confused me

**Review by arl — COMMENTED** [review]
Why is the `httpload` package under `url_shortener`? They should both be considered different binaries.
IMHO the directory structure of the whole repo should be made to stick to that of the generally accepted convention, which is:

```
repo:
 - ./cmd/programa
 - ./cmd/programb
 - ./pkg/foo
 - ./pkg/bar
 - go.mod
```

File: url_shortener/cmd/httpload/main.go [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+
+	return makeLoadRequest(ctx, url, results)
+}
+
+func main() {
+	flag.Parse()
+
+	nonFlagArgs := flag.Args()
+	if len(nonFlagArgs) != 1 {
+		fmt.Println(usageMessage)
```
[the remark this answers is not in the document]
Comment (rgianassi) [reply url_shortener/cmd/httpload/main.go]: I tried:

```go
	var flags = flag.NewFlagSet(os.Args[0], flag.ExitOnError)
	flags.SetOutput(nil)
	flags.Usage = func() {
		fmt.Fprintf(flags.Output(), "Usage of %s:\n", os.Args[0])
		flags.PrintDefaults()
	}

	config := newConfigFromFlags(flags)

	if err := config.parse(flags); err != nil {
		fmt.Println("main: error during arguments parsing. Error:", err)
		flags.PrintDefaults()
		os.Exit(exitCodeError)
	}
```

but nothing is printed with 2 arguments.

File: url_shortener/cmd/httpload/main.go (hunk 3 above)
Comment (rgianassi) [reply url_shortener/cmd/httpload/main.go]: I followed the documentation at: https://blog.golang.org/context/google/google.go

What code do you suggest to write?

File: httpload/cmd/httpload/main.go [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Total:        %12.4f secs", totalTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Slowest:      %12.4f secs", maxTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Fastest:      %12.4f secs", minTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Average:      %12.4f secs", averageTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Requests/sec: %12.4f", requestsPerSecond))
+}
+
+func main() {
+	flags := flag.NewFlagSet(os.Args[0], flag.ContinueOnError)
```
Comment (arl) [inline httpload/cmd/httpload/main.go:210]: At the moment, calling `httpload -h` has not the expected effect.
Just changing the constant to `flag.ExitOnError` would solve that

**Review by arl — APPROVED** [review]
Apart from the -h comment, rest LGTM.
I'm waiting for the rest now ;-)

File: httpload/cmd/httpload/main.go [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Slowest:      %12.4f secs", maxTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Fastest:      %12.4f secs", minTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Average:      %12.4f secs", averageTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Requests/sec: %12.4f", requestsPerSecond))
+}
+
+func setupGracefulShutdown(done chan bool) {
+	signalChannel := make(chan os.Signal, 1)
+
+	signal.Notify(signalChannel, syscall.SIGINT, syscall.SIGKILL)
```
Comment (arl) [inline httpload/cmd/httpload/main.go:380]: Theoretically syscall package should never be used when we're dealing with multiplatform code.
you can use os.Interrupr and os.Kill, see https://godoc.org/os#Signal
At the moment they're aliased of syscall.SIGINT and SIGKILL but who knows it could change in the future

File: httpload/cmd/httpload/main.go [hunk 9]
```diff
@@ -0,0 +1,440 @@
...
+
+	if err := config.checkFlags(); err != nil {
+		log.Println("main: error checking flags. Error:", err)
+		return exitCodeError
+	}
+
+	loadTester := newLoadTesterFromConfig(config)
+
+	go func() {
+		err := loadTester.run(config, done)
```
Comment (arl) [inline httpload/cmd/httpload/main.go:408]: config has already been provided to `loadTester` at construction so why should we pass it again when calling run?

File: httpload/cmd/httpload/main.go [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+			return
+		}
+
+		if err != nil {
+			log.Println("main: error during load test. Error:", err)
+			os.Exit(exitCodeError)
+		}
+	}()
+
+	<-done
```
Comment (arl) [inline httpload/cmd/httpload/main.go:420]: We're starting a goroutine and immediately after, we wait for it to be terminated before going on with the program, which means this code could be simpler if `run` was a blocking call.
Generally, it's better to write, reason about and maintain if functions/methods are blocking. This has also the advantage that concurrency constructs such as channels do not leak a function/method scope.

For example: 

```go
...
loadTester := newLoadTesterFromConfig(config)
err := loadTester.run()
if err != nil {
    // handle actual error
    ...   
}
...
```

File: httpload/cmd/httpload/main.go [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+		log.Println("main: error checking flags. Error:", err)
+		return exitCodeError
+	}
+
+	loadTester := newLoadTesterFromConfig(config)
+
+	go func() {
+		err := loadTester.run(config, done)
+
+		if errors.Is(err, context.DeadlineExceeded) {
```
Comment (arl) [inline httpload/cmd/httpload/main.go:410]: `loadTester.run()` is leaking its implementation details here (the fact that `run` internally uses a context).
IMHO `run` should have itself deal with `context.DeadlineExceeded` and not require `trueMain` to know about it (i.e returns a nil error since it's not really an error)

File: httpload/cmd/httpload/main.go [hunk 12]
```diff
@@ -0,0 +1,440 @@
...
+	"net/http"
+	"os"
+	"os/signal"
+	"sort"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/pkg/errors"
```
Comment (arl) [inline httpload/cmd/httpload/main.go:18]: I'm not sure you really want to use this package. The go standard library now has added to https://godoc.org/errors many of the features of the package you imported

File: httpload/cmd/httpload/main.go [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+}
+
+func (lt *LoadTester) genResults(ctx context.Context, in <-chan Result) (<-chan error, error) {
+	errc := make(chan error, 1) // signals goroutine completion
+
+	go func() {
+		defer close(errc)
+
+		for result := range in {
+			lt.results = append(lt.results, result)
```
Comment (arl) [inline httpload/cmd/httpload/main.go:220]: race condition : you append to the results slice from multiple goroutines at once

**Review by arl — COMMENTED** [review]
It would be great to add some comments on the function that deal with concurrency, overall when they're part of a overall flow of elements (genRequests, genLoadRequests, etc.).

What I'd say about the code is that it shows that you quickly and clearly assimilated the concepts of channels and goroutines. I really like that the whole program builds a pipelines of requests, errors and ties all together and works. :-)

However, the code is a also bit hard to follow since there are a lot (too many?) channels involved, they're a bit overused. It's something everybody myself included does when they start go, they tend to overuse channels and goroutines. They are powerful constructs but they make the program quite complex to follow when they start to be passed around.  It already shows with 3 or 4 concurrent functions so I let you imagine in a large code base.

What I'd suggest would be to split the complexity into different struct with different responsibilities, each of which is self contained and can even be tested independently from the rest.

For example:
 - the loadTester.run() methods starts a goroutine which reads from a channel of results and stores them in a slice.
 - refactor some code in order to have a `worker` that is responsible to perform N requests and only N. Each worker runs independently from the others, maybe even having a start/stop methods of their own. At creation, a worker receives the results channel.
 - create N of those workers in the loadTester.run().
 - let the workers finish by themselves or call stop on each of them if a timeout was given

File: httpload/cmd/httpload/main.go (hunk 10 above)
Comment (rgianassi) [reply httpload/cmd/httpload/main.go:420]: In reality, I can come out with a message on `done` channel in case of shutdown, either. If the call is blocking, I cannot shutdown properly (unless Go magically handles it in a way I do not know). Or do you mean something else?

**Comment by rgianassi** [comment]
> * refactor some code in order to have a `worker` that is responsible to perform N requests and only N. Each worker runs independently from the others, maybe even having a start/stop methods of their own. At creation, a worker receives the results channel.

If a worker is slower than others, it slows down all other workers. I think if the N workers work in parallel reading from the queue as soon as they finish their job is more efficient and mantains the load at maximum. What do you think?
For the start/stop part: isn't it the `Context` responsibility?

File: httpload/cmd/httpload/main.go (hunk 11 above)
Comment (rgianassi) [reply httpload/cmd/httpload/main.go:411]: How do you suggest to solve the problem? Do I have to write a `run`/`trueRun` couple of methods? The alternative is to check for `DeadlineExceeded` on all `run`'s `ifs.

File: httpload/cmd/httpload/main.go (hunk 13 above)
Comment (rgianassi) [reply httpload/cmd/httpload/main.go:221]: It is only 1 goroutine reading from the `in` channel. I added a `mutex`, but isn't it the channel purposes to avoid using synchronization primitives? Am i missing something?

File: httpload/cmd/httpload/main.go (hunk 11 above)
Comment (rgianassi) [reply httpload/cmd/httpload/main.go]: Do not mind it, maybe I solved.

File: httpload/cmd/httpload/main.go (hunk 10 above)
Comment (rgianassi) [reply httpload/cmd/httpload/main.go]: Maybe, I can call the channel `shutdown` or something similar to clarify its function. What do you think?

File: httpload/cmd/httpload/main.go (hunk 12 above)
Comment (rgianassi) [reply httpload/cmd/httpload/main.go]: Maybe it can help: the `noise` is generated by the `make test`  command, but I do not know why.

File: httpload/cmd/httpload/main.go (hunk 10 above)
Comment (arl) [reply httpload/cmd/httpload/main.go]: The graceful shutdown was not a requirement for this application, I think it's safe to assume that if you CTRL-C while the loadtester is running you're not interested by the probably incomplete results you'll get. I mean it can  be an improvement sure, but let's add it further.

If we want to keep it, we could remove the done channel and use a context that manages the whole application lifetime.
It gets cancelled on CTRL-C. Also if the config specified a timeout, you create a child context out of it (using `context.WithTimeout(parent, timeout)`.

Anyway, about making the RunLoaderPipeline method blocking, I don't see what you mean, the code is already blocking since the execution stops just after calling RunLoaderPipeline and waits for the done channel.

File: httpload/cmd/httpload/main.go (hunk 12 above)
Comment (arl) [reply httpload/cmd/httpload/main.go]: Just write the import yourself. replace `github.com/pkg/errors` with just `errrors`.
`gofmt` only adds/removes package but never overwrites one so when it's confused, just write it yourself

File: httpload/cmd/httpload/main.go (hunk 13 above)
Comment (arl) [reply httpload/cmd/httpload/main.go]: sorry my bad I didn't express myself correctly, the data race is not due to multiple concurrent writes but concurrent read/writes, specifically:

```
==================
WARNING: DATA RACE
Read at 0x00c00010dbc0 by main goroutine:
  github.com/rgianassi/learning/go/httpload/pkg/loader.(*LoadTester).DumpStatuses()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/pkg/loader/loadtester.go:40 +0x108
  github.com/rgianassi/learning/go/httpload/pkg/loader.(*LoadTester).WriteResults()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/pkg/loader/loadtester.go:109 +0x155
  main.trueMain()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/cmd/httpload/main.go:61 +0x4ba
  main.main()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/cmd/httpload/main.go:75 +0x17a

Previous write at 0x00c00010dbc0 by goroutine 15:
  github.com/rgianassi/learning/go/httpload/pkg/loader.(*LoadTester).AppendResults.func1()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/pkg/loader/loadtester.go:208 +0x13a

Goroutine 15 (finished) created at:
  github.com/rgianassi/learning/go/httpload/pkg/loader.(*LoadTester).AppendResults()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/pkg/loader/loadtester.go:204 +0x7d
  github.com/rgianassi/learning/go/httpload/pkg/loader.(*LoadTester).RunLoaderPipeline()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/pkg/loader/loadtester.go:294 +0x3b3
  main.trueMain.func1()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/cmd/httpload/main.go:50 +0x46
```

Comment (arl) [reply httpload/cmd/httpload/main.go]: `loadtester.go:40` and `loadtester.go:208` are called at the same time from 2 different goroutines, hence the data race

Comment (rgianassi) [reply httpload/cmd/httpload/main.go]: Maybe, I am not understanding something obvious, but `loadtester.go:208` is called inside `AppendResults` which is a stage of the pipeline in `RunLoaderPipeline` which blocks until the pipeline is ended (because of `<-done` statement). On the other end, `loadtester.go:40` reads when the pipeline is done, so no worker is writing to it anymore (and it must not, otherwise we have partial results). How do you get the data race condition? I tried, but I am not able to.

**Comment by arl** [comment]
> > * refactor some code in order to have a `worker` that is responsible to perform N requests and only N. Each worker runs independently from the others, maybe even having a start/stop methods of their own. At creation, a worker receives the results channel.

> If a worker is slower than others, it slows down all other workers. I think if the N workers work in parallel reading from the queue as soon as they finish their job is more efficient and mantains the load at maximum. What do you think?

I agree that in other context the fact that a worker works faster than another could be a problem.
However in the context of our load-tester I'm not sure it matters, we're speaking about milliseconds here.

I'd also add that the concurrency pattern required to distribute work over multiple workers is slightly more involved and a bit harder to get it right that the one I proposed, less efficient but simpler.

I mean that's a good optimisation we can do later but IMHO we should focus on getting things right.

> For the start/stop part: isn't it the `Context` responsibility?

I don't want to answer with a non-answer, but you can yourself decide where to set the bar of th e context responsibility :-) 
There are multiple ways to achieve the same goal but what we should focus one:
 - synchronous API (and concurrent internals)
 - clear separation of concerns
 - readable, maintainable code
